### PR TITLE
feat(source): onboard satismeter as an event-stream

### PR DESCRIPTION
**Pull request by anantjain45823**
## Description of the change

> Onboarding new Source SatisMeter supporting Track Event using WebHook to receive data.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [x] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/sources/satismeter/utils.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/* eslint-disable no-param-reassign */
+const { isDefinedAndNotNullAndNotEmpty } = require("../../util");
+
+/**
+ * This function removes all those variables which are
+ * empty or undefined or null from all levels of object.
+ * @param {*} obj
+ * @returns payload without empty null or undefined variables
+ */
+const refinePayload = obj => {
```
Comment (yashasvibajpai) [inline v0/sources/satismeter/utils.js:10]: Is there any scope of making this available in root-level utils, so that it can be reusable?

(Non blocking)

Comment (anantjain45823) [reply v0/sources/satismeter/utils.js:10]: Yeah absolutely we can do it .. but I guess there were changes going on regarding the is isdefinednotnullandnotempty one. Will reach out respective stakeholders and will do the appropriate.

File: v0/sources/satismeter/transform.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  if (!message?.userId && !message?.anonymousId) {
+    message.anonymousId = event?.response.Id
+      ? event.response.id
+      : md5(event.response.user.email);
+  }
+  return message;
+}
+
+function process(event) {
+  // console.log("input: ", event);
```
Comment (ItsSudip) [inline v0/sources/satismeter/transform.js:28]:
```suggestion
```

File: v0/sources/satismeter/transform.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  const response = processEvent(event);
+  const returnValue = refinePayload(response);
+  // console.log("output: ", returnValue);
```
Comment (ItsSudip) [inline v0/sources/satismeter/transform.js:31]:
```suggestion
```

File: v0/sources/satismeter/transform.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+const buildTrackPayload = event => {
+  const message = new Message(`SATISMETER`);
+  message.setEventType("track");
+  message.setPropertiesV2(event, trackMapping);
+  message.event = `Survey ${event.event}`;
+  console.log("location:", event.response.location);
+  return message;
+};
+
+function processEvent(event) {
```
Comment (ItsSudip) [inline v0/sources/satismeter/transform.js:16]:
```suggestion
const processEvent = event => {
```

File: v0/sources/satismeter/transform.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  // making sure we are returning either userId or AnonymousId
+  if (!message?.userId && !message?.anonymousId) {
+    message.anonymousId = event?.response.Id
+      ? event.response.id
+      : md5(event.response.user.email);
+  }
+  return message;
+}
+
+function process(event) {
```
Comment (ItsSudip) [inline v0/sources/satismeter/transform.js:27]:
```suggestion
const process = event => {
```

File: v0/sources/satismeter/transform.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  return message;
+};
+
+function processEvent(event) {
+  const message = buildTrackPayload(event);
+  // making sure we are returning either userId or AnonymousId
+  if (!message?.userId && !message?.anonymousId) {
+    message.anonymousId = event?.response.Id
+      ? event.response.id
+      : md5(event.response.user.email);
```
Comment (ItsSudip) [inline v0/sources/satismeter/transform.js:22]: What if the email is also not there?
```suggestion
      : md5(event.response?.user?.email);
```

File: v0/sources/satismeter/transform.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+const { refinePayload } = require("./utils");
+const trackMapping = require("./data/trackMapping.json");
+
+// Ref: https://support.gainsight.com/PX/Integrations/01Technology_Partner_Integrations/Integrate_with_Gainsight_PX_Using_Webhooks
+const buildTrackPayload = event => {
+  const message = new Message(`SATISMETER`);
+  message.setEventType("track");
+  message.setPropertiesV2(event, trackMapping);
+  message.event = `Survey ${event.event}`;
+  console.log("location:", event.response.location);
```
Comment (ItsSudip) [inline v0/sources/satismeter/transform.js:12]:
```suggestion
```

File: __tests__/data/satismeter_source.json [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+                      "value": [
+                        "a1",
+                        "b1"
+                      ],
+                      "type": "multiple-choice"
+                    }
+                  ],
+                "surveyDismissed":false,
+                "surveyPending": false,
+                "receivedAt":"2022-11-17T17:51:21.764Z"
```
Comment (ItsSudip) [inline __tests__/data/satismeter_source.json:226]: Can't we put `receivedAt` to root?

File: v0/sources/satismeter/transform.js (hunk 6 above)
Comment (anantjain45823) [reply v0/sources/satismeter/transform.js:22]: I guess then we can remove this as anonymousId is anyhow present at all times

Comment (anantjain45823) [reply v0/sources/satismeter/transform.js:22]: And as a fallback, we can be dependent t response.id as it acts as an identifier for a response

File: __tests__/data/satismeter_source.json (hunk 8 above)
Comment (anantjain45823) [reply __tests__/data/satismeter_source.json:226]: `receivedAt` at the root level mainly is regarding when the event was received by the rudder-server and what we want to convey is when the response was recorded by Satismeter.
Ref: https://www.rudderstack.com/docs/event-spec/standard-events/common-fields/#:~:text=client%20to%20RudderStack.-,receivedAt,-The%20timestamp%20when

File: v0/sources/satismeter/transform.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  // making sure we are returning either userId or AnonymousId
+  if (!message?.userId && !message?.anonymousId) {
+    message.anonymousId = event?.response.id; // fallback to response.id as response.id is always present in payload
+  }
+  return message;
+};
+
+const process = event => {
+  const response = processEvent(event);
+  const returnValue = refinePayload(response);
```
Comment (yashasvibajpai) [inline v0/sources/satismeter/transform.js:24]: Can we have a more informative name here in place of return value, something related to the response or payload.

Comment (anantjain45823) [reply v0/sources/satismeter/transform.js]: Done

File: v0/sources/satismeter/transform.js (hunk 6 above)
Comment (ItsSudip) [reply v0/sources/satismeter/transform.js]: Okay

File: v0/util/index.js [hunk 10]
```diff
@@ -1793,6 +1793,36 @@ const getDestAuthCacheInstance = destType => {
...
+      typeof obj[ele] === "object" &&
+      !Array.isArray(obj[ele])
+    ) {
+      const refinedObject = refinePayload(obj[ele]);
+      if (Object.keys(refinedObject).length !== 0) {
+        refinedPayload[ele] = refinePayload(obj[ele]);
+      }
+    } else if (
+      typeof obj[ele] === "boolean" ||
+      // eslint-disable-next-line no-restricted-globals
```
Comment (ItsSudip) [inline v0/util/index.js:1816]: can't we remove this restriction?

Comment (anantjain45823) [reply v0/util/index.js:1816]: No, we can't as `isDefinedAndNotNullAndNotEmpty` does not work on boolean and numbers thats why we have to take them into consideration.

Comment (anantjain45823) [reply v0/util/index.js]: Done

**Review by ItsSudip — APPROVED** [review]

**Review by yashasvibajpai — APPROVED** [review]

**Pull request merged**